open_spiel/games/battleship.h

```diff
@@ -0,0 +1,358 @@
+// `damage_pl1 - loss_multiplier * damage_pl2`, where `loss_multiplier` is a
+// parameter to the generator.
+//
+// When `loss_multiplier = 1`, the game is zero-sum.
+// Note that currently no partial credit is awarded to a player that hit but did
+// not sink a ship.
+//
+//
+// Parameters
+//     "board_width"           int     Number of columns of the game board for
+//                                     each player            (default = 10)
+//     "board_height"          int     Number of rows of the game board for
+//                                     each player            (default = 10)
+//     "ship_sizes"          [int]     Length of the ships each player has
+//                                                      (default = [2,3,3,4,5])
+//     "ship_values"      [double]     Value of the ships each player has
+//                                                      (default = [1,1,1,1,1])
+//     "num_shots"             int     Number of shots available to each
+//                                     player                 (default = 50)
+//     "allow_repeated_shots" bool     If false, the players will be prevented
+//                                     from shooting multiple times at the same
+//                                     cell of the board      (default = true)
+//     "loss_multiplier"    double     Loss multiplier (see above). The game is
+//                                     zero-sum iff the loss multiplier is 1.0
+//                                                            (default = 1.0)
+//
+// NOTE: The list parameters must be supplied as a string of semicolon-separated
+//       values, wrapped in square brackets. For example: "[1;2]" is a list with
+//       elements `1` and `2`. "[1]" is a list with only one element.
+//
+// References
+// ==========
+//
+// [1]:
+// https://papers.nips.cc/paper/9122-correlation-in-extensive-form-games-saddle-point-formulation-and-benchmarks.pdf
+//
+// If you want to reference the paper that introduced the benchmark game, here
+// is a Bibtex citation:
+//
+// ```
+// @inproceedings{Farina19:Correlation,
+//    title=    {Correlation in Extensive-Form Games: Saddle-Point Formulation
+//               and Benchmarks},
+//    author=   {Farina, Gabriele and Ling, Chun Kai and Fang, Fei and
+//               Sandholm, Tuomas},
+//    booktitle={Conference on Neural Information Processing Systems
+//               (NeurIPS)},
+//    year={2019}
+// }
+// ```
+
+#ifndef OPEN_SPIEL_GAMES_BATTLESHIP_H_
+#define OPEN_SPIEL_GAMES_BATTLESHIP_H_
+
+#include <memory>
+#include <vector>
+
+#include "open_spiel/games/battleship_types.h"
+#include "open_spiel/spiel.h"
+
+namespace open_spiel {
+namespace battleship {
+
+constexpr int kDefaultBoardWidth = 10;
+constexpr int kDefaultBoardHeight = 10;
+constexpr const char* kDefaultShipSizes = "[2;3;3;4;5]";
+constexpr const char* kDefaultShipValues = "[1.0;1.0;1.0;1.0;1.0]";
```

Comment:
Can you make these std::string_view?

Reply:
Also, please make all of the constants `inline constexpr T`: https://abseil.io/tips/168.

Reply:
1. Ignore me- I forgot about the `GameParameter` c'tor. It looks like it's non-trivial to change, so leave it as-is. 

2. `absl::string_view` is preferable, I missed that you were already using it.